Mingw64 build fixes

cam888eron opened the pull request:
On mingw64, openssl build hangs in mkdef.pl. Apparent endless
loop. Root cause found to be some header file declarations like:
  void foo();
instead of
  void foo(void);
Main perl loop which gets rid of argument list demands at least 1 character between the parentheses.
The fix gets rid of empty argument list '()' before existing main loop. Lets build succeed on mingw64.

The mingw64 ld apparently does not export symbols while linking a shared library. Only get
~half dozen symbols in ssl and crypto .dll's. Expected 100's of symbols. The change makes it happen.

Environment:    Windows 10    msys2 of 20161025    mingw64 latest as of JUN/2017
gcc version 6.3.0 (Rev3, Built by MSYS2 project)
perl 5, version 24, subversion 1 (v5.24.1) built for x86_64-msys-thread-multi

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte commented:
> The mingw64 ld apparently does not export symbols while linking a shared library. Only get
~half dozen symbols in ssl and crypto .dll's. Expected 100's of symbols. The change makes it happen.

Didn't you also get a `libcrypto.dll.a` and `libssl.dll.a`?  They are import libraries produced by the same ld, have you looking in them?

levitte commented on util/mkdef.pl:
```diff
@@ -874,6 +874,10 @@ sub do_defs
+			}
 			# fold round brackets recursively: (t(*v)(t),t) -> (t{}{},t) -> {}
 			while(/\(.*\)/s) {
 				s/\([^\(\)]+\)/\{\}/gs;
```
Wouldn't it be easier to change the `+` to a `*` on this line?

cam888eron replied:
Re the .a static libraries, I looked at them briefly. After the fix in this PR, libcrypto.dll.a contains ~4600 entries like d004039.o . Presumably .a is correct. (But I should have said 'Expected 1000's of symbols' earlier.)

Apparently source file name gets lost along the way on mingw64. I am used to *NIX, where foo.c results in foo.o showing in the .a file.

cam888eron replied:
Re perl code, I did not understand what you meant. I hacked the perl. I do not know perl, so I hacked it based on similarity to other languages which I do know, C, bash script, Ruby, etc. I stopped hacking the perl when opensll build succeeded.

If a perl expert can find a better fix, please do by all means...

cam888eron replied:
Correction to my last post, I think I figured out what you meant re perl code. Trying now, takes a few hours for full build...

cam888eron replied:
You are correct, your idea works. May have found the perl expert I was hoping for, thanks.

levitte commented on Configurations/10-main.conf:
```diff
@@ -1498,7 +1498,7 @@ sub vms_info {
         shared_target    => "mingw-shared",
         shared_cflag     => add("-D_WINDLL"),
-        shared_ldflag    => "-static-libgcc",
+        shared_ldflag    => "-static-libgcc -Wl,--export-all",
```
You should remove this.  I'll explain in another comment

levitte commented:
> Re the .a static libraries, I looked at them briefly. After the fix in this PR, libcrypto.dll.a contains ~4600 entries like d004039.o . Presumably .a is correct. (But I should have said 'Expected 1000's of symbols' earlier.)

Ok, so here's what seems to be a misunderstanding...  the way things work on Windows (at least with our builds), and that includes mingw, is that when you build a shared library, you *also* build a so called import library, which is really just a transfer vector that (roughly speaking) maps symbol to specific locations in the shared library. The import library file name may look like a static library, but it's not the same thing as the actual static library.

So, with mingw, you end up with this:

static library: libcrypto.a, libssl.a
shared library: libcrypto-1_1.dll, libssl-1_1.dll
import library: libcrypto.dll.a, libssl.dll.a

This is, as far as I understand, supported naming in msys/mingw, i.e. when ld receives the option `-lcrypto`, it will first look for `libcrypto.dll.a`, then `libcrypto.a` (unless `-static` is given, then only the latter is looked for).

richsalz commented:
CLA on file, thanks.

levitte commented:
> CLA on file, thanks.

Or mark the commit as trivial by adding the following separate line to its commit message:

```
CLA: trivial
```

For formality, I would say this is a trivial fix.

cam888eron commented:
Reverted export symbols. Thanks for pointing out that I overlooked significance of  'import lib'.

cam888eron commented:
Looks like CLA will not clear. Is that a typo 4 lines above? My email address, no dot between gmail and com. Not sure if my typo. If so, howto fix?

richsalz commented:
Not your fault, don’t worry about it.

levitte commented:
> Looks like CLA will not clear. Is that a typo in cla-check field on this GUI? My email address, no dot between gmail and com. Not sure if my typo. If so, howto fix?

That's in your git configuration, and *is* an incorrect address.  You can look for yourself like this:

```
git config --show-origin --get user.email
```

cam888eron commented:
Fixed email address at my end, thanks for tip. Please try to clear cla-check 1 more time.

levitte commented:
Ok cool.  However, it won't show up here unless you modify your commits.  Have a look [here](https://help.github.com/articles/changing-author-info/) for the github tips on how to do that.

cam888eron commented:
Ran script suggested by levitte, fixed my email address. cla-check is clear now. But git wants to redo a great many commits to rewrite history. I wonder if it would be better to delete this branch and redo pull request under different branch name? What do you think?

richsalz commented:
Rebase against master and ```git push -f``` should work.

FdaSilvaYY commented:
Rebase command :  `git pull --no-commit --rebase openssl master -v`
Push : `git push origin HEAD -fv`

mattcaswell commented:
Not sure what you did, but your rebase did not work properly. This PR is now showing over 250 commits.

cam888eron commented:
Modified FdaSilvaYY command slightly, since my:

  $ git remote -v # says:
  origin  https://github.com/cam888eron/openssl.git (fetch)
  origin  https://github.com/cam888eron/openssl.git (push)
  upstream        https://github.com/openssl/openssl.git (fetch)
  upstream        https://github.com/openssl/openssl.git (push)

so I used:

  $ git pull --no-commit --rebase upstream master -v # says:
---snip---  <overran terminal with 1000's of lines re mode like:>
   delete mode 100644 util/with_fallback.pm
  First, rewinding head to replay your work on top of it...
  Applying: BC-32.pl: refresh Borland C support.
  error: Failed to merge in the changes.
  Using index info to reconstruct a base tree...
  A       util/pl/BC-32.pl
  Falling back to patching base and 3-way merge...
  CONFLICT (modify/delete): util/pl/BC-32.pl deleted in HEAD and modified in BC-32.pl: refresh Borland C support.. Version BC-32.pl: refresh Borland C support. of util/pl/BC-32.pl left in tree.
  Patch failed at 0001 BC-32.pl: refresh Borland C support.
  The copy of the patch that failed is found in: .git/rebase-apply/patch

  When you have resolved this problem, run "git rebase --continue".
  If you prefer to skip this patch, run "git rebase --skip" instead.
  To check out the original branch and stop rebasing, run "git rebase --abort".

I tried further git rebase with --continue, --skip, and --abort; no joy.

My offer still stands. I am willing to delete anything/everything to
do with me from OpenSSL github, then recreate from scratch. Recreating
PR should work well after that. Not a big deal to me. What do you think?

levitte commented:
The times things get screwed up like that for me, I usually throw away the checkout and start over...
